test/parsing/vass/bang-vs-new-etc.chpl

```diff
@@ -0,0 +1,209 @@
+// This is a study of what our parser does at the moment.
+
+// The .good contains a few lines where "expected" and "actual" are the same
+// string. Once #13946 is fixed, these lines should disappear from the output.
+
+class C { }
+class D { type t; }
+proc tm() type return C;
+proc tn(type arg) type return D(arg);
+type CN = C?;
+
+param Cm = "C.meth";
+proc C.meth param  return Cm;
+proc type C.tmeth type  return D(int);
+proc type (shared C).tmeth type  return D(real);
+class X { var fld; }
+proc type C.xmeth type  return X;
+
+param Dtm = "D.tdmeth";
+proc type D.tdmeth param  return Dtm;
+
+var xx = new C();
+var yy: owned C?;
+var dd = new D(int);
+
+proc check(param name, type tp, param answer) {
+  if tp:string == answer then /*ok*/;
+  else compilerWarning(name, "  expected: ", answer, "  actual: ", tp:string);
+}
+
+proc check(param name, const ref arg, param answer) {
+  if (arg.type):string == answer then /*ok*/;
+  else compilerWarning(name, "  expected: ", answer, "  actual: ", (arg.type):string);
+}
+
+proc checkCm(param name, param answer) {
+  if answer == Cm then /*ok*/;
+  else compilerWarning(name, "  expected: ", Cm, "  actual: ", answer);
+}
+
+proc checkDtm(param name, param answer) {
+  if answer == Dtm then /*ok*/;
+  else compilerWarning(name, "  expected: ", Dtm, "  actual: ", answer);
+}
+
+proc main {
+  var z01a = new C()?!;
+  check("z01a = new C()?!", z01a, "borrowed C");
+
+  var z01b = new owned C()?!;
+  check("z01b = new owned C()?!", z01b, "borrowed C"); // get owned C
+
+  var z01c = new unmanaged C()?!;  // error: duplicate decorators
+  check("z01c = new unmanaged C()?!", z01c, "unmanaged C");
+
+  var z02 = new C?()!;
+  check("z02 = new C?()!", z02, "borrowed C");
+
+  // var z = new C()!?;  -- error "? on a value"
+  // var z = new C()??; -- syntax error
```

Comment:
I think both of these have fairly well-defined meanings according to the proposed precedence table.

`new C()!?` returns the same as `new borrowed C?()`
`new C()??` is the same as `new C?()`

For another PR